c2rust-analyze/src/dataflow/mod.rs

```diff
@@ -373,6 +421,9 @@ trait PropagateRules<T> {
     ) -> (T, T);
     fn all_perms(&mut self, ptr: PointerId, perms: PermissionSet, val: &T) -> T;
     fn no_perms(&mut self, ptr: PointerId, perms: PermissionSet, val: &T) -> T;
+    /// Apply a filter to restrict updates.  The result is similar to `new`, but all flags marked
+    /// in `updates_forbidden` are adjusted to match their `old` values.
+    fn restrict_updates(&mut self, old: &T, new: &T, updates_forbidden: &T) -> T;
```

Comment:
Are both implementations of this identical? If so, it might make sense to just provide the implementation here.

Reply:
That would require a bunch of constraints on `T` in the trait definition.  (The constraints are already satisfied by both impls, but would be annoying to write down.)